### PR TITLE
[TACHYON-434] s3 client unit tests

### PR DESCRIPTION
**Pull request by calvinjia**
https://tachyon.atlassian.net/browse/TACHYON-434

This PR makes the string manipulations methods and adds unit tests for the S3UnderFileSystem.

### CODE REVIEW

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+public class S3UnderFileSystemFactoryTest {
+  @Test
+  public void factoryTest() {
+    TachyonConf conf = new TachyonConf();
+
+    UnderFileSystemFactory factory =
+        UnderFileSystemRegistry.find("s3n://localhost/test/path", conf);
+    Assert.assertNotNull(
+        "A UnderFileSystemFactory should exist for s3n paths when using this module", factory);
+  }
```
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java]: Shall we also test the negative case: passing an illegal path prefix (e.g. "foo://", "s3n//", "s3n:/") and expect null?

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java [hunk 2]
```diff
@@ -16,23 +16,131 @@
...
+    String result2 = Whitebox.invokeMethod(sMockS3UnderFileSystem, "getParentKey", input2);
+    String result3 = Whitebox.invokeMethod(sMockS3UnderFileSystem, "getParentKey", input3);
+    String result4 = Whitebox.invokeMethod(sMockS3UnderFileSystem, "getParentKey", input4);
+
+    Assert.assertEquals("s3n://test-bucket", result1);
+    Assert.assertEquals(null, result2);
+    Assert.assertEquals("s3n://test-bucket/parent", result3);
+    Assert.assertEquals(null, result4);
+  }
+
```
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java]: How about adding isFolderTest?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 3]
```diff
@@ -347,6 +333,23 @@ private boolean deleteInternal(String key) {
   }
 
   /**
+   * Gets the child name based on the parent name. If the parent is not valid, the entire child
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: "If parent is not valid, the entire child key will be returned."
If parent is not valid (i.e., not the prefix of child), isn't this an error case in our code path?
If true, we shouldn't simpley return the original child string (with a warning though). Instead, we should escalate the error to the caller. e.g., by returning "".

Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: I have similar concerns on stripPrefixIfPresent as I think if the cases when s3 prefix doesn't present, it is an error case. 

Though, stripFolderSuffixIfPresent seems totally legitimate to me, as we may see folder or filer which has no suffix.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 4]
```diff
@@ -483,6 +486,33 @@ private boolean mkdirsInternal(String key) {
...
+   * Treating S3 as a file system, checks if the parent directory exists.
+   * @param key the key to check
+   * @return true if the parent exists or if the key is root, false otherwise
+   */
+  private boolean parentExists(String key) {
+    // Assume root always has a parent
+    if (isRoot(key)) {
+      return true;
+    }
+    String parentKey = getParentKey(key);
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: shouldn't we check if parentKey equals null?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java [hunk 5]
```diff
@@ -371,6 +374,10 @@ private StorageObject getObjectDetails(String key) {
    * @return the the parent key
    */
   private String getParentKey(String key) {
+    // Root does not have a parent.
```
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: NIT: Usually, arranging multiple `if-else` in the following pattern makes src code much easier to follow (for a new/average audience):
```
if (foo) {
  return false;
}
if (bar) {
  return false;
}
return SomeFunc();
```

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Also we shall unit-test `parentExists`

**Comment by apc999** [comment]
Thanks for adding the unit-tests, these are really helpful. I put some comments. PTAL

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java (hunk 2 above)
Comment (calvinjia) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java]: Currently, isFolder requires a check to S3.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (calvinjia) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Current, `parentExists` also requires an S3 backend.

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java (hunk 1 above)
Comment (calvinjia) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java]: I'm usually against negative tests because there's technically the space of all possible strings, but perhaps one negative case is sufficient to ensure that the factory isn't just returning true for everything. What do you think?

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 3 above)
Comment (calvinjia) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: I agree about getChildName.
The naming of `stripPrefixIfPresent` seems to imply that its okay for the prefix not to be present. Currently its optional because we will call it on user inputs which may not contain the prefix (for example they may pass in paths relative to the bucket which should be allowed).

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java (hunk 1 above)
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemFactoryTest.java]: yes, adding one negative case is sufficient IMO, just to prevent the factory returning true for all inputs, as you mentioned.

File: underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java (hunk 2 above)
Comment (apc999) [inline underfs/s3/src/test/java/tachyon/underfs/s3/S3UnderFileSystemTest.java]: Oh, that's right. Perhaps in future PRs, we could considering injecting mock S3. But in this PR, I think it is good :)

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 3 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Oh I see. then probably you shall update its javadoc accordingly as this is not very obvious.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: Right. Thanks

**Comment by calvinjia** [comment]
@apc999 Thanks for the feedback, I've updated the code accordingly.

File: underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java]: I still think we should probably check the returned value of `parentKey` from line 495. It looks to me possible that `parentKey = null` in line 495 ( e.g., when `parentExists("somKeyFromUser")` was called from mkdirs). In this case, calling `isFolder(null)` on line 496 will throw an exception.

There are also places that we should check returned value from `getParent`, e.g.,  line 86, 231, where returned null from `getParent` may bring down the function.

**Comment by apc999** [comment]
Only one comment left regarding `parentKey`. Either if it is fixed, or it turns out my analysis goes wrong for the potential exception, then I will LGTM this PR :) Thanks a bunch on pushing the tests.

**Comment by calvinjia** [comment]
@apc999 sorry I missed that comment, do you think its properly addressed now?

**Comment by apc999** [comment]
LGTM

Thanks!
